Update project build configuration and dependencies

dzmpr opened the pull request:
Hi there!
I have some troubles with m2 bottom sheet navigator so I decided to add m3 bottom sheet navigator and solve this problem for me and for others. But when I started working on it I saw that voyager uses compose multiplatform 1.6.11 when I needed 1.7. One after another I'm ended up made some improvement to project build configuration and updated all dependencies (not only compose kmp).
So, what's have been done:
* Updated gradle and dependencies.
* Cleaned up project from redundant `.gitignore` and `customer.proguard.pro` files.
* Got rid of `buildSrc` and replace it with `build-config` included build. With `buildSrc` there are problems with using version catalogs and no pros.
* Split modules configuration to pluggable convention plugins to make configuration simpler and cleaner.
* Removed some unused or redundant dependencies (such as `appcompat`, `compose-runtimeSaveable`, `compose-compiler`, `compose-animations`, and maybe something else). Note: this compose dependencies are not multiplatform and only used in samples.
* Replaced `com.benasher44:uuid` dependency with kotlin stdlib implementation which is available from kotlin 2.0.
* Replaced `compileOnly` for compose dependencies with plain `implementation` as now seems all working fine but with `compileOnly` compiler threw warnings about js targets.
* Fixed js sample and some more minor fixes for samples.

What's I left behind:
* I can't figure out custom target hierarchy that voyager uses, so left it as is (with `jvm("desktop")` target which triggers compiler warning).
* Didn't update api dump, because I removed `ExperimentalVoyagerApi` from `nonPublicMarkers` (I consider this is a mistake, because it' actually public api, just experimental). So if you are okay with this I update api with additional commit.

I've tested all samples on android and following targets of multiplatform sample: js, wasm, macos (not sure what target).

So I'm waiting for review, and open for any questions if you have one.

dzmpr committed:
Update gradle 8.2 => 8.11.1

dzmpr committed:
Remove redundant .gitignore's; use recommended rules from github

dzmpr committed:
Remove empty consumer rule files

dzmpr committed:
Remove app compat dependency as only theme were used from it

dzmpr committed:
Update project dependencies; simplify build configuration; move configuration logic to convention plugins

dzmpr committed:
Update yarn.lock; fix js sample

dzmpr committed:
Use implementation instead of compileOnly for compose dependencies

dzmpr committed:
Remove uuid dependency and use kotlin stdlib implementation

dzmpr committed:
Replace deprecated API's in samples

dzmpr committed:
Minor warning fixes

Vodes commented on gradle/libs.versions.toml:
```diff
@@ -1,77 +1,65 @@
+mavenPublish = "0.30.0"
+composeMultiplatform = "1.7.1"
+binaryCompatibilityValidator = "0.16.3"
+atomicfu = "0.26.1"
```
Fyi, this will currently cause an issue because compose expects (and forces) 0.23.2.
Seems to be a known issue https://youtrack.jetbrains.com/issue/CMP-5831

dzmpr replied:
I think it's compose multiplatform problem rather than voyager. Anyway I tested voyager with cmp 1.8.0-rc01 and it seems all works fine.
I intended to update this PR when cmp 1.8.0 will be released.

dzmpr committed:
Rename common jvm group from jvm to commonJvm to make it non ambiguous

Syer10 commented:
> @Vodes why did you change from 8.2 => 8.11.1 It is to new mine one still using 8.7.2

Because its good practice to always update dependencies, Gradle is always meant to be up-to-date for the improvements in speed and bugfixes.

hafiz013 commented:
@Syer10 has this fork or pr merge to master repo veyoger?

dzmpr committed:
Disable androidResources in library modules

dzmpr committed:
Remove package from AndroidManifest

dzmpr committed:
Update dependencies

dzmpr committed:
Replace deprecated LifecycleEffect

dzmpr committed:
Merge branch 'main' into dependencies_update

# Conflicts:
#	samples/multiplatform/build.gradle.kts

dzmpr commented:
Just updated this PR, what did I done more:
* Renamed `jvm` targets group to `commonJvm` to make it non-ambigious with `jvm` target.
* Disabled android resources in library modules as voyager doesn't use them. It should have positive impact on build speed.
* Removed `package` from AndroidManifest's as from AGP 8.0 namespace should be declared in module build.gradle file.
* Updated CMP to 1.8.0, Kotlin to 2.1.21, Gradle to 8.14, and other dependencies to the most actual versions. The only exception is koin as they mistakenly removed required targets from latest releases. So I'm waiting for [this issue](https://github.com/kosi-libs/Kodein/issues/481) to be done to update koin.

There is still some work to be done, e.g. update ktlint and resolve problems that it finds and update samples. But this is out of scope of this PR, so I intended to do this after this PR is merged.

@DevNatan, @DevSrSouza can you please take a look at this PR.

programadorthi commented on voyager-bottom-sheet-navigator/build.gradle.kts:
```diff
@@ -16,17 +15,11 @@ kotlin {
-        jvmTest.dependencies {
-            implementation(libs.junit.api)
-            runtimeOnly(libs.junit.engine)
+            implementation(compose.runtime)
```
Voyager uses `compileOnly` to avoid transitive dependency at compile time.

dzmpr replied:
Seems using `compileOnly` for compose dependencies break compilation for certain targets, here the warning I get:
```text
A compileOnly dependency is used in targets: Kotlin/JS, Kotlin/Native, Kotlin/Wasm.
Dependencies:
    - org.jetbrains.compose.runtime:runtime-saveable:1.8.1 (source sets: iosArm64Main, iosSimulatorArm64Main, iosX64Main, jsMain, macosArm64Main, macosX64Main, wasmJsMain)
    - org.jetbrains.compose.runtime:runtime:1.8.1 (source sets: iosArm64Main, iosSimulatorArm64Main, iosX64Main, jsMain, macosArm64Main, macosX64Main, wasmJsMain)

Using compileOnly dependencies in these targets is not currently supported, because compileOnly dependencies must be present during the compilation of projects that depend on this project.

To ensure consistent compilation behaviour, compileOnly dependencies should be exposed as api dependencies.
```

Furthermore, it is not possible to use voyager without using compose, why we should to avoid making a transitive dependency here? We not strictly enforce compose version so consumer just will use more recent version of compose.

I also checked out decompose for reference and it using implementation without any problems as it seems.

programadorthi commented on voyager-core/src/androidMain/kotlin/cafe/adriel/voyager/androidx/AndroidScreenLifecycleOwner.kt:
```diff
@@ -136,7 +136,7 @@ public class AndroidScreenLifecycleOwner private constructor() :
     private fun getHooks(): List<ProvidedValue<*>> {
         atomicAppContext.compareAndSet(null, LocalContext.current.applicationContext)
-        atomicParentLifecycleOwner.compareAndSet(null, LocalLifecycleOwner.current)
+        atomicParentLifecycleOwner.compareAndSet(null, LocalLifecycleOwner.current) // deprecated
```
Remove the comment

programadorthi reviewed (commented):
Can you share evidences of all samples working after the whole changes?

dzmpr commented on build-config/src/main/kotlin/configuration/configure-android.gradle.kts:
```diff
@@ -0,0 +1,37 @@
+
+        if (isAndroidLibraryModule()) {
+            val proguardFilename = "consumer-rules.pro"
+            if (layout.projectDirectory.file(proguardFilename).asFile.exists()) {
```
AGP will print that certain proguard files not found:

> Supplied consumer proguard configuration does not exist: /Users/dzmitry/AndroidStudioProjects/voyager/voyager-bottom-sheet-navigator/consumer-rules.pro

If you ok with this I remove an if clause.

programadorthi replied:
You removed the empty files. That is the reason to complain about missing it.
Is there any advantage to remove the files instead of putting the check for them?

dzmpr replied:
In my opinion a project has less clutter when it doesn't have files that don't affect anything.

dzmpr commented on build.gradle.kts:
```diff
@@ -1,44 +1,34 @@
+        add(projects.samples.multiModule.navigation)
+    }
+    with(nonPublicMarkers) {
+        add("cafe.adriel.voyager.core.annotation.InternalVoyagerApi")
```
Actually code marked with `ExperimentalVoyagerApi` is a part of public API surface of the voyager. So it shouldn't be treated as non-public and developer need to be aware about changes in this code.

hafizkloudius014 commented:
When can this will merge to master?

dzmpr commented on gradle/samples.versions.toml:
Can you please explain what conflicts do you mean? Samples modules are accessed through `projects.samples.*` while dependencies just with `samples.*`.
Maybe `libsSamples.*` or `samplesLibs.*` or `demo.*` will be a better naming?

I think it's good to have more fine-grained control over what dependencies you add to the library. Now it's not possible to 'accidentally' add new dependency from common catalog. You need to make a mindful decision to add it to to libs catalog and it's becomes more clear to the reviewer when new dependency were added to library.

programadorthi replied:
Maybe samplesCatalog? I'm not good with names. For the first read I made assumptions between `projects.samples` and `samples.*`. Then a I realized that was a catalog.

dzmpr commented on gradle/samples.versions.toml:
```diff
@@ -0,0 +1,14 @@
+[versions]
+leakCanary = "2.14"
+koin = "4.0.4"
+compose = "1.8.1"
```
This is compose which used by samples, so basically by the consumer of the voyager library. You have no control over what library version voyager consumers are using so I don't see a problem there.

dzmpr committed:
Remove unused repositories

dzmpr committed:
Update .gitignore

dzmpr commented on settings.gradle.kts:
```diff
@@ -1,19 +1,50 @@
         google()
         mavenCentral()
-        maven(url = "https://maven.pkg.jetbrains.space/public/p/compose/dev" )
+        gradlePluginPortal()
+        mavenLocal()
```
Agreed, I'll remove them

dzmpr commented on build-config/src/main/kotlin/configuration/configure-ktlint.gradle.kts:
```diff
@@ -0,0 +1,8 @@
+plugins.apply("org.jlleitschuh.gradle.ktlint")
+
+configure<KtlintExtension> {
+    version.set("0.47.1")
```
Yes I can with hack like [this](https://github.com/gradle/gradle/issues/15383#issuecomment-1567461389)

programadorthi replied:
Maybe [this](https://github.com/programadorthi/kotlin-routing/blob/504f636708ca093636289c0e0555ae3839018247/buildSrc/src/main/kotlin/JvmConfig.kt#L18)?

dzmpr committed:
Remove comment

programadorthi commented:
I think soon it'll be merged. It's a good contribution. Just some adjustments and done.

hafizkloudius014 commented:
@dzmpr i found bugs regarding to parseable serialize. 

error : when time you off the screen and come back to screen error about parse. so i come out this solution but don't know how to constribute:

```abstract class BasedActivityScreen : Screen, JavaSerializable{   //<-- without JavaSerializable will be error.
    @Composable
    override fun Content() {
        previewScreen()
    }

    @Preview
    @Composable
    open fun previewScreen(){}
}```

and i hope that 
```
public expect interface Screen { <-- can add parceable

    public open val key: ScreenKey

    @Composable
    public fun Content()
}

internal fun Screen.commonKeyGeneration() =
    this::class.multiplatformName ?: error("Default ScreenKey not found, please provide your own key")
```

programadorthi commented:
@hafizkloudius014 Screen is JavaSerializable by default. More details [here](https://voyager.adriel.cafe/state-restoration/).

hafizkloudius014 commented:
Is this pr good to go to merge?

dzmpr committed:
Get ktlint version from version catalog

dzmpr commented on samples/android/src/main/AndroidManifest.xml:
```diff
@@ -1,11 +1,10 @@
         android:name=".App"
         android:label="@string/app_name"
-        android:theme="@style/Theme.AppCompat.Light.NoActionBar"
+        android:theme="@android:style/Theme.Material.Light.NoActionBar"
```
It's a compose app which uses compose theme instead of xml one. So I removed appcompat dependency and used built-in theme.

dzmpr committed:
Update dependencies

dzmpr committed:
Rename samples catalog

dzmpr committed:
Rename jvm module convention plugin

programadorthi commented:
Close good. Now we need other maintainers approve. I'm talking with them to take a time to review.

dzmpr commented:
Glad to hear that! Here are screenshots from all targets where I know how to run KMP app:
<details><summary>iOS</summary>

</details>
<details><summary>iOS</summary>

</details>
<details><summary>iOS</summary>

</details>
<details><summary>iOS</summary>

</details>
* iOS
<img width="568" alt="multiplatform-ios" src="https://github.com/user-attachments/assets/821d588b-6106-485a-a86d-042826c2c7ea" />

* macOS
<img width="912" alt="multiplatform-macos" src="https://github.com/user-attachments/assets/1f10f884-74ca-4b1d-9c69-d769f2546edd" />

* wasmJs
<img width="835" alt="multiplatform-wasm" src="https://github.com/user-attachments/assets/675739f9-1785-48b4-b8ba-eb533e1d5c78" />

* js
<img width="795" alt="multiplatform-js" src="https://github.com/user-attachments/assets/43cef19a-ee38-48e3-8463-d81d91f86270" />

* android
![multiplatform-android](https://github.com/user-attachments/assets/e6c69b2e-02af-454f-b0b2-c33c56167eec)

dzmpr commented:
Glad to hear that! Here are screenshots from all target where I know how to run KMP app:
<details><summary>iOS</summary>
<img width="568" alt="multiplatform-ios" src="https://github.com/user-attachments/assets/44647a8a-002a-4e2f-b957-10100c88fba8" />
</details>
<details><summary>macOS</summary>
<img width="912" alt="multiplatform-macos" src="https://github.com/user-attachments/assets/eea703b4-c0fe-4ca0-92fd-130f32817be9" />
</details>
<details><summary>wasmJs</summary>
<img width="835" alt="multiplatform-wasm" src="https://github.com/user-attachments/assets/42e98d01-b58e-479f-918f-bbea16e7edcd" />
</details>
<details><summary>js</summary>
<img width="795" alt="multiplatform-js" src="https://github.com/user-attachments/assets/897ed68b-60a1-473c-afdc-b494088fe636" />
</details>
<details><summary>android</summary>
<img width="540" alt="multiplatform-android" src="https://github.com/user-attachments/assets/6e5e421d-3b91-4975-b8a3-dd0be7bbaa5f" />
</details>

programadorthi commented:
@dzmpr build is broken. I think some compose APIs are now JVM 11 and we need to upgrade the target too.

dzmpr commented:
@programadorthi Ok, I'll update JVM target.

If you agree with my decision to remove `ExperimentalVoyagerApi` from non-public API I'll update api dump too.

DevSrSouza commented:
Soon I will review it! Have not be able to do it still!

If there is no API Breaking changes, there is not need for bumping voyager to 2.0 even that we update Kotlin and Compose to new versions.

DevSrSouza commented:
@dzmpr the CI is falling to compile desktop, can you fix that?

DevSrSouza commented on build-config/src/main/kotlin/configuration/configure-kotlin-jvm.gradle.kts:
```diff
@@ -0,0 +1,33 @@
+                    it.target.targetName == "desktop" || it.target is KotlinAndroidTarget
+                }
+            }
+            group("nonAndroid") {
```
why does this group have spaces in it name?

DevSrSouza commented on build-config/src/main/kotlin/configuration/configure-kotlin-multiplatform.gradle.kts:
```diff
@@ -0,0 +1,60 @@
+                    it.target.targetName == "desktop" || it.target is KotlinAndroidTarget
+                }
+            }
+            group("nonAndroid") {
```
Same as here, why the spaces

DevSrSouza commented on build-config/src/main/kotlin/configuration/configure-opt-in.gradle.kts:
```diff
@@ -0,0 +1,10 @@
+configure<KotlinProjectExtension> {
+    sourceSets.all {
+        languageSettings {
+            optIn("cafe.adriel.voyager.core.annotation.InternalVoyagerApi")
+            optIn("cafe.adriel.voyager.core.annotation.ExperimentalVoyagerApi")
```
It has spaces here as well in the string, I don't think it should work

DevSrSouza replied:
Well ignore this, github rendering issue

DevSrSouza commented on build-config/src/main/kotlin/configuration/configure-publishing.gradle.kts:
```diff
@@ -0,0 +1,49 @@
+    publishToMavenCentral(host = SonatypeHost.S01, automaticRelease = true)
+    signAllPublications()
+
+    pom {
+        description = "A pragmatic navigation library for Jetpack Compose"
+        inceptionYear = "2021"
+        url = "https://github.com/adrielcafe/voyager"
+
+        licenses {
+            license {
+                name = "The MIT License"
+                url = "https://opensource.org/licenses/MIT"
+                distribution = "repo"
+            }
+        }
+
+        scm {
+            url = "https://github.com/adrielcafe/voyager"
+            connection = "scm:git:ssh://git@github.com/adrielcafe/voyager.git"
+            developerConnection = "scm:git:ssh://git@github.com/adrielcafe/voyager.git"
+        }
+
+        developers {
+            developer {
+                id = "adrielcafe"
+                name = "Adriel Cafe"
+                url = "https://github.com/adrielcafe/"
+            }
+        }
+    }
+}
```
Why this was required? why not keep it on `gradle.properties`?

dzmpr replied:
I think explicit configuration is better than implicit

DevSrSouza commented on build-config/src/main/kotlin/voyager-jvm-module.gradle.kts:
```diff
@@ -0,0 +1,4 @@
+plugins {
+    id("voyager-base-module")
```
space here in the string as well

DevSrSouza replied:
Well ignore this, github rendering issue

DevSrSouza commented on build-config/src/main/kotlin/voyager-kmp-module.gradle.kts:
```diff
@@ -0,0 +1,4 @@
+plugins {
+    id("voyager-base-module")
```
space here in the string as well

DevSrSouza replied:
Well ignore this, github rendering issue

DevSrSouza commented on gradle.properties:
```diff
@@ -1,54 +1,23 @@
-
-kotlin.mpp.androidSourceSetLayoutVersion=2
-
-# Maven
-SONATYPE_HOST=S01
-SONATYPE_AUTOMATIC_RELEASE=true
-RELEASE_SIGNING_ENABLED=true
-
-GROUP=cafe.adriel.voyager
-
-POM_DESCRIPTION=A pragmatic navigation library for Jetpack Compose
-POM_INCEPTION_YEAR=2021
-POM_URL=https://github.com/adrielcafe/voyager
-
-POM_LICENCE_NAME=The MIT License
-POM_LICENCE_URL=https://opensource.org/licenses/MIT
-POM_LICENCE_DIST=repo
-
-POM_SCM_URL=https://github.com/adrielcafe/voyager
-POM_SCM_CONNECTION=scm:git:ssh://git@github.com/adrielcafe/voyager.git
-POM_SCM_DEV_CONNECTION=scm:git:ssh://git@github.com/adrielcafe/voyager.git
-
-POM_DEVELOPER_ID=adrielcafe
-POM_DEVELOPER_NAME=Adriel Cafe
-POM_DEVELOPER_URL=https://github.com/adrielcafe/
```
Why have you move this to code? The plugin does not support gradle properties anymore?

DevSrSouza commented on gradle/libs.versions.toml:
```diff
@@ -1,77 +1,66 @@
-composeRuntime = "1.5.0"
-composeUI = "1.5.0"
-rxjava = "3.1.5"
+koin-compose = "4.0.4"
```
Coin version here has a space

DevSrSouza commented on gradle/libs.versions.toml:
```diff
@@ -1,77 +1,66 @@
+composeMultiplatform = "1.8.1"
+binaryCompatibilityValidator = "0.17.0"
+atomicfu = "0.27.0"
+ktlint = "0.47.1"
```
ktlint here has space

DevSrSouza replied:
Well ignore this, github rendering issue

DevSrSouza commented on settings.gradle.kts:
```diff
@@ -1,19 +1,48 @@
+        gradlePluginPortal()
+    }
+    resolutionStrategy {
+        eachPlugin {
+            if (requested.version == null) return@eachPlugin
+            when (requested.id.id) {
+                "com.android.tools.build" -> {
+                    useModule("${requested.id.id}:gradle:${requested.version}")
+                }
+            }
+        }
+    }
```
Any reason why this is needed?

dzmpr replied:
By default Gradle to resolve plugin artifact by plugin id uses plugin id as group id and add `.gradle.plugin` suffix to plugin id to construct artifact id. So plugin `"pluginId" version "1.0.0"` becomes `"pluginId:pluginId.gradle.plugin:1.0.0"` artifact. 
But AGP build plugin has custom artifact id which we need to override.

DevSrSouza commented on voyager-core/src/commonWebMain/kotlin/cafe.adriel.voyager.core/screen/ScreenKey.web.kt:
```diff
@@ -1,5 +1,7 @@
+import kotlin.uuid.Uuid
 
-internal actual fun randomUuid(): String = uuid4().toString()
+@OptIn(ExperimentalUuidApi::class)
```
Unnecessary spacing here.

DevSrSouza committed:
Fixing crash causes in older versions because of removeFirst and removeLast

DevSrSouza committed:
Bump JVM Target to 11 because of Compose requirements, fix Thread Safe collections compilation, fix Lifecycle KMP

DevSrSouza committed:
Fixing lint issues

dzmpr committed:
Update jvm target for samples

dzmpr committed:
Update dependencies

dzmpr committed:
Update voyager API dump

dzmpr commented:
@DevSrSouza Hey, did you get a chance to test the PR locally with `implementation` as you mentioned?

dzmpr committed:
Update dependencies; change publishing host to central portal

hafiz013 commented:
hi it is already july, it is ok can merge this pr? thanks you.

dzmpr committed:
Update API dump

devnatan reviewed (approved):
LGTM

hafiz013 commented:
nice it approved. may i know what is latest version of this library? thank you

zouyonghao commented:
Any updates?

dzmpr commented:
@DevSrSouza Hi, can you please take a look at this PR? There is one step left to release and unblock other improvements for the lib.

hafizkloudius014 commented:
hi @programadorthi  should be increase the number of version library before can merge to main?

Robotxm commented:
Hi @programadorthi, it seems that [publish of new version](https://github.com/adrielcafe/voyager/actions/runs/17022534639) is failed. Will it be fixed?

hafizkloudius014 commented:
hi @programadorthi  error:
```
Failed to stop service 'sonatype-repository-build-service'.
> Upload failed: {"error":{"message":"Invalid token"}}
```

dzmpr commented:
I think this https://github.com/adrielcafe/voyager/pull/514#issuecomment-3028613220 may be a problem.

hafizkloudius014 commented:
ok then their problem not migrate the portal. @dzmpr
